envoy/common/execution_context.h

```diff
@@ -65,4 +81,22 @@ class ScopedExecutionContext : NonCopyable {
   ExecutionContext* context_;
 };
 
+#ifdef ENVOY_ENABLE_EXECUTION_SCOPE
+#define ENVOY_EXECUTION_SCOPE_CAT_(a, b) a##b
+#define ENVOY_EXECUTION_SCOPE_CAT(a, b) ENVOY_EXECUTION_SCOPE_CAT_(a, b)
+// Invoked when |scopedObject| is active from the current line to the end of the current c++ scope.
+// |executionContext| is a pointer to the current ExecutionContext.
+// |scopedObject| is a pointer to a Envoy::Tracing::Span or a Http::FilterContext.
+#define ENVOY_EXECUTION_SCOPE(executionContext, scopedObject)                                      \
+  Envoy::Cleanup ENVOY_EXECUTION_SCOPE_CAT(on_scope_exit_, __LINE__) =                             \
+      [execution_context = (executionContext), scoped_object = (scopedObject)] {                   \
+        if (execution_context == nullptr || scoped_object == nullptr) {                            \
+          return Envoy::Cleanup::Noop();                                                           \
+        }                                                                                          \
+        return execution_context->onScopeEnter(*scoped_object);                                    \
+      }()
```

Comment:
Please take the connection directly. Then we needn't to implement the executionContext method in other classes. Let's try to aggregate all execution related logic here.

Reply:
I considered that, but I think it's better to use executionContext() instead of connection, mainly for maintainability.

There are currently 5 places where we override the `executionContext()` method, excluding test: see [link](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/envoyproxy/envoy%24+%22+executionContext%28%29+const%22+-file:test&patternType=keyword&case=yes&sm=0).

There are currently 33 places where a `ScopeTrackerScopeState` object is instantiated, excluding test: see [link](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/envoyproxy/envoy%24+%22ScopeTrackerScopeState+scope%22+OR+%22ScopeTrackerScopeState%3E+s%22+-file:test&patternType=keyword&sm=0).   My Envoy binary also adds some extra instances(based on ScopeTrackedObjectStack), when some async operation resumes.

If we get rid of the 5 `executionContext` methods, we'll need to add 33 `ScopedExecutionContext` objects, one per `ScopeTrackerScopeState`, to achieve feature parity.   Since `ScopeTrackerScopeState` appears per-function and `executionContext` appears per-class, it is much more likely to see new `ScopeTrackerScopeState` introduced in the future, compared to new classes that need a `executionContext` method.

Furthermore, the places where `ScopeTrackerScopeState` is instantiated have different ways to access to the connection, in some cases, like [this one in TimerImpl](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/event/timer_impl.cc?L22-24), there is currently no access to the connection, and adding access to the connection would feel unnatural.

Let's keep executionContext method as is, and only use macro for `ENVOY_EXECUTION_SCOPE`?

Reply:
**I think the bigger problem of the whole feature is that the ExecutionContext actually is an custom abstraction that used only by your private feature and are make no sense for other modules or users.** Except in your private fork, no one know what is the implementation looks like.

So, my only concern is not to hybrid the execution context related code/abstraction to other modules/code base and let it be a complete independent feature. 

Rather than introducing the ExecutionContext to the code base, I will prefer to add macro at where the observation is necessary.
Although it may introduce more changes, but actually is more cleaner because all other modules needn't to aware the specific abstraction. It's completed independent and easy to manage.

But as you said, there are some position the you cannot access the connection. A compromised way is to expose filter state or stream info by the `ScopeTrackedObject`, then you can access your private ExecutionContext from it.

Reply:
> **I think the bigger problem of the whole feature is that the ExecutionContext actually is an custom abstraction that used only by your private feature and are make no sense for other modules or users.** Except in your private fork, no one know what is the implementation looks like.
> 

I agree ExecutionContext is a abstraction that currently only makes sense in my Envoy server. However, I don't understand why it is a problem? For most users, they won't need to be aware of it and their Envoy will work perfectly fine.
If, in the unlikely case, some user wants to do something at the start and end of ScopeTrackerScopeState, they can implement their own ExecutionContext subclass and do whatever in the active/deactive methods. If it still doesn't work for some user, I think that's a good time for all the stakeholders to discuss a better way to do it. After all, we can only find "common" solutions if there are multiple use cases to begin with.

> So, my only concern is not to hybrid the execution context related code/abstraction to other modules/code base and let it be a complete independent feature.
> 
> Rather than introducing the ExecutionContext to the code base, I will prefer to add macro at where the observation is necessary. Although it may introduce more changes, but actually is more cleaner because all other modules needn't to aware the specific abstraction. It's completed independent and easy to manage.
> 

The coupling of ExecutionContext and ScopeTrackerScopeState is not accidental - My Envoy server needs to do something at the beginning and end of each ScopeTrackerScopeState, but ScopeTrackerScopeState doesn't provide a hook to do it, that's why ExecutionContext is introduced. Said another way, ExecutionContext is not a "standalone" feature, it is just a hook that gets invoked at ScopeTrackerScopeState's ctor and dtor.

> But as you said, there are some positions the you cannot access the connection. A compromised way is to expose filter state or stream info by the `ScopeTrackedObject`, then you can access your private ExecutionContext from it.

Reply:
> I agree ExecutionContext is a abstraction that currently only makes sense in my Envoy server. However, I don't understand why it is a problem? For most users, they won't need to be aware of it and their Envoy will work perfectly fine.

A private feature or abstraction should be maintained in the their private fork. Or every envoy distributor will add their custom logic in the open envoy. 
I don't think this is good way to maintain OSS except the feature is a general feature or be implemented as an extension.

The only reason about why we discuss the problem here is because that your feature require to penetrate to the core deeply and you don't want to maintain the custom patch at your private fork. Even at this way, I don't want to reject this PR directly, but at least we should try our best to make it clean and separate it with other core code.

Reply:
> The coupling of ExecutionContext and ScopeTrackerScopeState is not accidental - My Envoy server needs to do something at the beginning and end of each ScopeTrackerScopeState, but ScopeTrackerScopeState doesn't provide a hook to do it, that's why ExecutionContext is introduced. Said another way, ExecutionContext is not a "standalone" feature, it is just a hook that gets invoked at ScopeTrackerScopeState's ctor and dtor.

The doesn't change thing's nature? This hook still a private thing and lots of (some) core classes need to inherit from and implementation it.

Reply:
I don't waste your paying and time, but also cannot accept this in this way now.

So, how about this? I can creat a PR try to refactor it first, and you can as reviewer to review it to see is that acceptable to you?

Reply:
> I don't want to waste your paying and time, but also cannot accept this in this way now.
> 
> So, how about this? I can creat a PR try to refactor it first, and you can as reviewer to review it to see is that acceptable to you?

Sounds Good! That's probably the best way to resolve our deadlock:) Once you have a PR, I can try to integrate it into my server and see whether/how well it works.